common/persistence/sql/sqlExecutionManager.go

```diff
@@ -622,8 +624,217 @@ func updateBufferedEvents(tx sqldb.Tx, batch *p.DataBlob, clear bool, shardID in
+		newRunID := sqldb.MustParseUUID(insertExecutionInfo.RunID)
+
+		startVersion := common.EmptyVersion
+		lastWriteVersion := common.EmptyVersion
+		if insertReplicationState != nil {
+			startVersion = insertReplicationState.StartVersion
+			lastWriteVersion = insertReplicationState.LastWriteVersion
+		}
+
+		// 1. update current execution
+		// TODO: https://github.com/uber/cadence/issues/1679
+		if err := updateCurrentExecution(tx,
+			shardID,
+			domainID,
+			insertExecutionInfo.WorkflowID,
+			sqldb.MustParseUUID(insertExecutionInfo.RunID),
+			insertExecutionInfo.CreateRequestID,
+			insertExecutionInfo.State,
+			insertExecutionInfo.CloseStatus,
+			startVersion,
+			lastWriteVersion,
+		); err != nil {
+			return &workflow.InternalServiceError{
+				Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed at updateCurrentExecution. Error: %v", err),
+			}
+		}
+
+		// 2. lock base run: we want to grab a read-lock for base run to prevent race condition
+		// It is only needed when base run is not current run. Because we will obtain a lock on current run anyway.
+		if request.BaseRunID != currExecutionInfo.RunID {
+			filter := &sqldb.ExecutionsFilter{ShardID: shardID, DomainID: domainID, WorkflowID: workflowID, RunID: sqldb.MustParseUUID(request.BaseRunID)}
+			_, err := tx.ReadLockExecutions(filter)
+			if err != nil {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed at ReadLockExecutions. Error: %v", err),
+				}
+			}
+		}
+
+		// 3. update or lock current run
+		if request.UpdateCurr {
+			if err := createTransferTasks(tx, request.CurrTransferTasks, shardID, domainID, workflowID, currRunID); err != nil {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed to create transfer tasks. Error: %v", err),
+				}
+			}
+			if err := createTimerTasks(tx, request.CurrTimerTasks, nil, shardID, domainID, workflowID, currRunID); err != nil {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed to create timer tasks. Error: %v", err),
+				}
+			}
+
+			if err := lockAndCheckNextEventID(tx, shardID, domainID, workflowID, currRunID, request.Condition); err != nil {
+				switch err.(type) {
+				case *p.ConditionFailedError:
+					return err
+				default:
+					return &workflow.InternalServiceError{
+						Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed to lock executions row. Error: %v", err),
+					}
+				}
+			}
+
+			if err := updateExecution(tx, currExecutionInfo, currReplicationState, shardID); err != nil {
+				return &workflow.InternalServiceError{
+					Message: fmt.Sprintf("ResetWorkflowExecution operation failed. Failed to update executions row. Erorr: %v", err),
+				}
+			}
+		} else {
+			if err := lockAndCheckNextEventID(tx, shardID, domainID, workflowID, currRunID, request.Condition); err != nil {
```

Comment:
can you add some comments on why these locks need to be acquired / what is base/current/updateCurr etc ?

Reply:
I already verified and I failed the test when I forgot to add those fields